### PR TITLE
add hover, sheet parser

### PR DESCRIPTION
**Pull request by pmlich**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -96,24 +189,119 @@ var app = new Vue({
...
+                    if (app.form.attributes[i].sum == AttrSums[0] && !(changed.includes(i))) {
+                        app.form.attributes[i].allowed = AttrAllowed[0]
                         AttrSums.shift()
                         AttrAllowed.shift()
                         changed.push(i)
                     }
                 }
             }  
-        }
+        },
```
Comment (pmlich) [inline index.js:210]: dalej to funguje tak, ze 
- sheet najprv prekreslim do blank sheetu, kde prepisem iba existujuce hodnoty. Spravim si ale kopiu blanku cez cloneArray.
- ulozi sa to do blank premenej
- potom to parsujem do struktury, vegenerujem ju podla data.form_structure

a potom by sa malo spravit prepocitanie sum+priorit, to tu nemam, lebo pouzivam iny kod. Tak to niekto dorobte. Je tam zakomentovana funkcia.

File: index.html [hunk 2]
```diff
@@ -4,24 +4,56 @@
...
-<script src="https://cdn.jsdelivr.net/npm/vue@2/dist/vue.js"></script>
-
 <body>
+
 <main id="chargen">
+
     <h1>CharGen</h1>
-    <div>
+
+    <div id="Info" class="clear">
```
Comment (haterzlin) [inline index.html:13]: I was thinking if we need id here. Like id="Info" - I suppose it seems to be useful to have it for future and also it's logical to have it, but do we actually use it?

Same for other cases bellow, Attributes, Skills and other

File: index.html [hunk 3]
```diff
@@ -4,24 +4,56 @@
...
+        <h2>Info</h2>
+        <div is="info-category"
+	     v-for="(item, index) in form.info"
+             v-bind:group_id="'info'"
```
Comment (haterzlin) [inline index.html:17]: why there are two types of quotes? 
' and "

File: index.html [hunk 4]
```diff
@@ -4,24 +4,56 @@
...
 <body>
+
 <main id="chargen">
+
     <h1>CharGen</h1>
-    <div>
+
+    <div id="Info" class="clear">
+        <h2>Info</h2>
+        <div is="info-category"
```
Comment (haterzlin) [inline index.html:15]: why rather <div is="info-category" instead of <info-category ?

File: index.html [hunk 5]
```diff
@@ -4,24 +4,56 @@
...
+	     v-for="(item, index) in form.info"
```
Comment (haterzlin) [inline index.html:16]: why we specify index twice? First here in for and then on lie 19 again?

File: index.html [hunk 6]
```diff
@@ -4,24 +4,56 @@
...
-            v-bind:category="item"
-            v-bind:key="item.id"
-            v-on:sumchanged="RecomputeAttributePriorities()"
-        >
-        </attribute-category>
+        <div is="attribute-category"
+	     v-for="(item, index) in form.attributes"
+             v-bind:category="item"
+             v-on:sumchanged="RecomputeAttributePriorities()"
+	></div>
```
Comment (haterzlin) [inline index.html:30]: - [ ] fix padding

File: index.html [hunk 7]
```diff
@@ -4,24 +4,56 @@
...
+	></div>    
+
     </div>
+
+    <div id="skills" class="clear">
+        <h2>Skills</h2>
+        <div is="attribute-category"
+	     v-for="(item, index) in form.skills"
+             v-bind:category="item"
+             v-on:sumchanged="RecomputeAttributePriorities()"
```
Comment (haterzlin) [inline index.html:39]: - [ ] fix padding

File: index.html [hunk 8]
```diff
@@ -4,24 +4,56 @@
...
+	></div>    
+    </div>
+
+    <div id="spells" class="clear">
+        <h2>Spells</h2>
+        <div is="attribute-category"
```
Comment (haterzlin) [inline index.html:45]: - [ ] fix padding

File: index.html [hunk 9]
```diff
@@ -4,24 +4,56 @@
...
+	     v-for="(item, index) in form.spells"
+             v-bind:category="item"
+             v-on:sumchanged="RecomputeAttributePriorities()"
+	></div>    
+    </div>
+
+
 </main>
+
+<script src="vue.min.js"></script><!--script src="https://cdn.jsdelivr.net/npm/vue@2/dist/vue.js"></script-->
```
Comment (haterzlin) [inline index.html:55]: - [ ] use vue.js development version

File: index.html [hunk 10]
```diff
@@ -4,24 +4,56 @@
...
+             v-on:sumchanged="RecomputeAttributePriorities()"
+	></div>    
+    </div>
+
+    <div id="spells" class="clear">
+        <h2>Spells</h2>
+        <div is="attribute-category"
+	     v-for="(item, index) in form.spells"
+             v-bind:category="item"
+             v-on:sumchanged="RecomputeAttributePriorities()"
```
Comment (haterzlin) [inline index.html:48]: if we use this more then once, how many of categories are computed after one attribute/skill change ?

File: index.js [hunk 11]
```diff
@@ -1,3 +1,51 @@
+const isExist = (ref)       => typeof ref != 'undefined' && ref!=null;
+const isArray = (ref)       => typeof ref=='object' && ref!=null;
+const isFunction = (ref)    => typeof ref=='function' && ref!=null;
+const isObject = (ref)      => typeof ref=='object' && ref!=null;
+const getValue = (ref, def) => isExist(ref) ? ref : def;
+
+function cloneObject(obj)
+	{
+	return JSON.parse(JSON.stringify(obj));	// nepouzivat Object.assign({}, obj), nefunguje spravne u parseru!
```
Comment (haterzlin) [inline index.js:9]: comment should be in english

File: index.js [hunk 12]
```diff
@@ -1,3 +1,51 @@
+const isExist = (ref)       => typeof ref != 'undefined' && ref!=null;
+const isArray = (ref)       => typeof ref=='object' && ref!=null;
+const isFunction = (ref)    => typeof ref=='function' && ref!=null;
+const isObject = (ref)      => typeof ref=='object' && ref!=null;
+const getValue = (ref, def) => isExist(ref) ? ref : def;
+
+function cloneObject(obj)
+	{
```
Comment (haterzlin) [inline index.js:8]: brackets here are used differently than in other code

- [ ] either use same notation as is used in surrounding code or rewrite surrounding code to use your annotation

File: index.js [hunk 13]
```diff
@@ -1,3 +1,51 @@
+const isExist = (ref)       => typeof ref != 'undefined' && ref!=null;
```
[the remark this answers is not in the document]
Comment (haterzlin) [reply index.js:1]: - [ ] please remove what is not used

File: index.js (hunk 1 above)
Comment (haterzlin) [reply index.js:210]: function cloneArray is not used anywhere in the code, there is just definition

File: index.js [hunk 14]
```diff
@@ -1,3 +1,51 @@
...
+    data: function() {
+           return  {
+	      for_id: 'field[' + this.category_id + '][' + this.attr.id + ']',
+	   }},
+    template: `
+    <tr>
+        <td><label for="{{for_id}}">{{ attr.id }}</label></td>
+        <td class="input"><input type="text" id="{{for_id}}" name="{{for_id}}"
+		v-bind:value="attr.value"
+                v-on:change=""
```
Comment (haterzlin) [inline index.js:44]: what's this?

**Review by haterzlin — CHANGES_REQUESTED** [review]
This needs lots of additional work and I strongly suggest to break it to smaller changes instead of this

**Review by pmlich — COMMENTED** [review]
- pridany app-data.js
- pridany vue.min.js ... ale, lubo to okomentoval, ze pouzivame develop verziu. Mne ide o to, aby ta verzia sa netahala z externeho webu, takze to tam niekto nahrajte a opravte html
- upravene css (proti main su tam nejake class navyse)
- index.html - trochu som to popresuval a upravil
- index.js - tam som...
--- pridal hover, upravil value, pridal kontrolu value na true/false + range
--- pridal funkciu isExist
--- opravil 'var' u recomputeSum a priority
--- prepisal app.data na funkciu a prepojil to na stahovanie configu (z app-data.js)
Takze, to by mali byt take drobne zmeny.

File: index.css [hunk 15]
```diff
@@ -3,9 +3,21 @@ html, body {
     padding: 0;
 }
 
-h2 {margin:0;}
-h3 {margin:0;}
-.clear {clear:both; float:none;}
+h1 {
```
Comment (haterzlin) [inline index.css:6]: h1, h2, h3 {margin: 0; } ?

File: index.js [hunk 16]
```diff
@@ -1,50 +1,4 @@
...
-    template: `
-    <tr>
-        <td><label for="{{for_id}}">{{ attr.id }}</label></td>
-        <td class="input"><input type="text" id="{{for_id}}" name="{{for_id}}"
-		v-bind:value="attr.value"
-                v-on:change=""
-        ></td>
-    </tr>`
-  });
+const isExist = (ref) => typeof ref != 'undefined' && ref!=null;
```
Comment (haterzlin) [inline index.js:1]: slo by som pridat komentar co ta "funkcia" robi?

File: index.js [hunk 17]
```diff
@@ -36,57 +38,92 @@ Vue.component('attribute-category', {
...
+            attr.value = attr.value==index ? index-1 : index
+            if (isExist(attr.value_range))
+                {
+                attr.value = attr.value<attr.value_range[0] ? attr.value_range[0] : attr.value>attr.value_range[1] ? attr.value_range[1] : attr.value
+                }
+        },
+        setClass: function(value, index, hover) { return {
+	    'circle'     : true,
+	    'circleBlack': index<=value,
+	    'circleGreen': hover>=0 && hover >value && index>value  && index<=hover,
```
Comment (haterzlin) [inline index.js:55]: všetko je šedé tak Green a Red nemá zmysel

File: index.js [hunk 18]
```diff
@@ -36,57 +38,92 @@ Vue.component('attribute-category', {
...
     props: ['attr'],
+    data: function() { return {
+	hover: -1
+    } },
+    methods: {
+        setAttrValue: function(attr, index) {
+            attr.value = attr.value==index ? index-1 : index
+            if (isExist(attr.value_range))
+                {
+                attr.value = attr.value<attr.value_range[0] ? attr.value_range[0] : attr.value>attr.value_range[1] ? attr.value_range[1] : attr.value
```
Comment (haterzlin) [inline index.js:49]: pridat komentar co to robi?

File: index.js [hunk 19]
```diff
@@ -36,57 +38,92 @@ Vue.component('attribute-category', {
...
+                        { id: "Intelligence", value: 1 },
+                        { id: "Wits", value: 1 },
+                        { id: "Resolve", value: 1 } 
+                    ],
+                    sum: 3,
+                    allowed: 5
+                },            
+            ]
+        }
+	data.props = config; // load from config (global variable in file 'app-data.js')
```
Comment (haterzlin) [inline index.js:121]: toto tu teraz nemusi byt

**Pull request merged**